### PR TITLE
Implemented leapfrog integrator

### PR DESCRIPTION
**Pull request by a-b-h-a-y-s-h-i-n-d-e**
## Description

added leapfrog integrator that supports both hamiltonian and non-hamiltonian dynamics.

## Type of change

Please check the option that is related to your PR.

- [  ] Bug fix (non-breaking change which fixes an issue)
- [ x ] New feature (non-breaking change which adds functionality)
- [  ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ x ] Documentations (modification for documents)

## Checklist

- [ x ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ x ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ x ] Run `mypy -p deepchem` and check no errors
  - [ x ] Run `flake8 <modified file> --count` and check no errors
  - [ x ] Run `python -m doctest <modified file>` and check no errors
- [ x ] I have performed a self-review of my own code
- [ x ] I have commented my code, particularly in hard-to-understand areas
- [ x ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ x ] New unit tests pass locally with my changes
- [ x ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import torch
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:1]: Can you please:

- Write unit-tests
- Add this to the utils.rst docs
- Add this function to __init__.py file of differentiation utils.

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+from torch.autograd import grad
+from tqdm import tqdm
+from typing import Callable
+
+
+
+def leapfrog(p: torch.Tensor,
+             q: torch.Tensor,
+             Func: Callable,
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: can we use lowercase for `Func`?

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+             dt: float,
+             N: int,
+             is_hamiltonian: bool = True) -> torch.Tensor:
+    """
+    Leapfrog integrator for simulating Hamiltonian or general dynamics.
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:15]: Please add a brief description of this integrator. Can you please add a reference?

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    p : torch.Tensor
+        Initial momentum tensor, shape (1,)
+    q : torch.Tensor
+        Initial position tensor, shape (1,)
+    Func : callable
+        If is_hamiltonian=True: Function that returns Hamiltonian H(z) where z=[p,q]
+        If is_hamiltonian=False: FuncCallabletion that returns derivatives [dp/dt, dq/dt]
+    dt : float
+        Time step size for integration
+    N : int
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: Please add `is_hamiltonian` to the docstring.

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
+import torch
+from torch.autograd import grad
+from tqdm import tqdm
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: Can you please avoid using `tqdm`, mainly because this doesn't match with the rest of deepchem's utils, classes, models, etc.

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    trajectories = torch.empty((N, 2))
+
+    p = p.clone().requires_grad_(True)
+    q = q.clone().requires_grad_(True)
+
+    if is_hamiltonian:
+        hamilt = Func(torch.cat([p, q], dim=-1))
+        dpdt = -grad(hamilt.sum(), q)[0]
+
+        for i in tqdm(range(N)):
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: I believe these `if`-`else` blocks are very similar, and we can generalize the code.

**Comment by JoseAntonioSiguenza** [comment]
Can you please fix formatting (`yapf`)?

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 5 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: done

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 1 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:1]: question on `utils.rst`

the rk methods are added under `Density Functional Theory Utilities`, now leapfrog also comes category of "integrators" so where to add this? or should I create new category for this ?

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 2 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: I have changed the `Func` to `fcn` which is used in rk methods too

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 6 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: the formula used is same for both the condition but the way we calculate gradients is different, just follow the `hamilt` variable in both the conditions

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 4 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py]: done

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 1 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:1]: quick update : added test case

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 3 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:15]: done

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    .. [4] https://github.com/zhengdao-chen/SRNN
+
+    """
+
+    trajectories = torch.empty((N, 2))
+
+    p = p.clone().requires_grad_(True)
+    q = q.clone().requires_grad_(True)
+
+    if is_hamiltonian:
```
Comment (rbharath) [inline deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:77]: Add some comments here inline on how the algorithm works

**Review by rbharath — COMMENTED** [review]
This PR is probably fine to merge once @JoseAntonioSiguenza 's comments are addressed, but in the next PR please implement a recurrent HNN and an explicit SRNN model (even if that model is just recurrent HNN + leapfrog)

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 7 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:77]: I have simplified the variable names as per the variables used in algorithm

e.g `p_half = p + (dt / 2) * dpdt` I guess this should be fine
because even if I add an inline comment it will be similar like this

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 1 above)
Comment (JoseAntonioSiguenza) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:1]: > question on `utils.rst`
> 
> the rk methods are added under `Density Functional Theory Utilities`, now leapfrog also comes category of "integrators" so where to add this? or should I create new category for this ?

@a-b-h-a-y-s-h-i-n-d-e I think under `Density Functional Theory Utilities` should be fine below current `deepchem.utils.differentiation_utils.integrate` functions.

File: deepchem/utils/test/test_differentiation_utils.py [hunk 8]
```diff
@@ -1293,3 +1293,21 @@ def fun(x):
     x1 = gd(fun, x0, [], terminate=True)
     x2 = gd(fun, x0, [], terminate=False)
     assert not torch.allclose(x1, x2)
+
+
+@pytest.mark.torch
+def test_leapfrog():
+    from deepchem.utils.differentiation_utils import leapfrog
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/test/test_differentiation_utils.py:1303]: Can you please add docstrings about what you aim to evaluate?

File: deepchem/utils/test/test_differentiation_utils.py [hunk 9]
```diff
@@ -1293,3 +1293,21 @@ def fun(x):
     x1 = gd(fun, x0, [], terminate=True)
     x2 = gd(fun, x0, [], terminate=False)
     assert not torch.allclose(x1, x2)
+
+
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/test/test_differentiation_utils.py:1297]: Can you please add for tests for leapfrog integrator? For example, can you please test energy conservation? Can you test leapfrog's results when `is_hamiltonian=False`? Can you test if leapfrog is symplectic?

**Comment by JoseAntonioSiguenza** [comment]
A PR was merged earlier today that should fix unit-test for `ubuntu-latest, 3.11 and 3.10`. Can you please rebase this PR and add the tests and docs suggestions? Also, I noticed multiple memory failures causing tests to fail. I believe they are not related to this PR, but a rebase will trigger the CI to run again.

File: deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py (hunk 1 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/differentiation_utils/integrate/symplectic_integrator.py:1]: done!

File: deepchem/utils/test/test_differentiation_utils.py (hunk 9 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/test/test_differentiation_utils.py:1297]: done!

File: deepchem/utils/test/test_differentiation_utils.py (hunk 8 above)
Comment (a-b-h-a-y-s-h-i-n-d-e) [reply deepchem/utils/test/test_differentiation_utils.py:1303]: done!

**Comment by JoseAntonioSiguenza** [comment]
LGTM, I'm concerned about `docs-build(3.9)` failure but I dont think is related to this PR.
